File: album-auction/src/app/schema.ts
import { PublicKey } from '@solana/web3.js';
import * as borsh from 'borsh';


/**
 * Album
 */
export class Album {
    id = 0;
    title = "";
    artist = "";
    winner = "";
    constructor(
        fields: {
            id: number,
            title: string,
            artist: string,
            winner: string,
        } | undefined = undefined
    )
    {
        if (fields) {
            this.id = fields.id;
            this.title = fields.title;
            this.artist = fields.artist;
            this.winner = fields.winner;
        }
    }
}

export const AlbumSchema = new Map([
    [Album, {
        kind: 'struct', fields: [
            ['id', 'u32'],
            ['title', 'u32'],
            ['artist', 'u32'],
            ['winner', 'u32'],
        ]
    }],
]);
  
export const ALBUM_SIZE = borsh.serialize(
    AlbumSchema,
    new Album(),
).length;


/**
 * Auction Instruction
 */
export enum AuctionInstructionCommand {
    BID,
    RESET,
}

export class AuctionInstruction {
    command = AuctionInstructionCommand.BID;
    localPubkey = PublicKey.default;
    constructor(
        fields: {
            command: AuctionInstructionCommand,
            localPubkey: PublicKey,
        } | undefined = undefined
    )
    {
        if (fields) {
            this.command = fields.command;
            this.localPubkey = this.localPubkey;
        }
    }
}

export const AuctionInstructionSchema = new Map([
    [AuctionInstruction, {
        kind: 'struct', fields: [['command', 'u32']]
    }],
]);
  
export const AUCTION_INSTRUCTION_SIZE = borsh.serialize(
    AuctionInstructionSchema,
    new AuctionInstruction(),
).length;

File: album-auction/src/auction_program/src/schema.rs
use {
    borsh::{
        BorshDeserialize,
        BorshSerialize
    },
    solana_program::{
        pubkey::Pubkey,
        program_error::ProgramError,
    },
    std::collections::HashMap,
};



/*
 * Album
 */
#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub struct Album {
    pub id: u8,
    pub title: String,
    pub artist: String,
    pub winner: Pubkey,
}


impl Album {

    pub fn match_album(
        account_key: &Pubkey, 
        program_id: &Pubkey,
        local_pubkey: &Pubkey,
    ) -> Result<Album, Error> {
        
        let ALBUM_ACCOUNT_SEEDS = HashMap::from([
            (String::from("beatles"), 
                Album {
                    id: 1, 
                    title: String::from("Abbey Road"),
                    artist: String::from("The Beatles"),
                    winner: program_id.clone(),
                }
            ),
            (String::from("queen"), 
                Album {
                    id: 2, 
                    title: String::from("News of the World"),
                    artist: String::from("Queen"),
                    winner: program_id.clone(),
                }
            ),
            (String::from("halen"),
                Album {
                    id: 3, 
                    title: String::from("Van Halen"),
                    artist: String::from("Van Halen"),
                    winner: program_id.clone(),
                }
            ),
            (String::from("acdc"),
                Album {
                    id: 4, 
                    title: String::from("Back in Black"),
                    artist: String::from("AC/DC"),
                    winner: program_id.clone(),
                }
            ),
        ]);
        
        for seed in ALBUM_ACCOUNT_SEEDS.keys() {
            if Pubkey::create_with_seed(
                local_pubkey,
                seed,
                program_id,
            )?.eq(account_key) {
                return match ALBUM_ACCOUNT_SEEDS.get(seed) {
                    Some(album) => album.clone(),
                    None => Err(ProgramError::InvalidInstructionData),
                }
            }
        }
        return Err(ProgramError::InvalidInstructionData)
    }
}
    

/**
 * Auction Instruction
 */
 #[derive(BorshSerialize, BorshDeserialize, Debug)]
pub enum AuctionInstructionCommand {
    BID,
    RESET,
}

#[derive(BorshSerialize, BorshDeserialize, Debug)]
pub struct AuctionInstruction {
    pub command: AuctionInstructionCommand,
    pub local_pubkey: Pubkey,
}